server/src/main/java/org/candlepin/controller/ActivationKeyController.java

```diff
@@ -0,0 +1,238 @@
+import org.candlepin.model.Owner;
+import org.candlepin.model.OwnerProductCurator;
+import org.candlepin.model.Pool;
+import org.candlepin.model.Product;
+import org.candlepin.model.Release;
+import org.candlepin.model.activationkeys.ActivationKey;
+import org.candlepin.model.activationkeys.ActivationKeyCurator;
+import org.candlepin.model.activationkeys.ActivationKeyPool;
+import org.candlepin.policy.js.activationkey.ActivationKeyRules;
+import org.candlepin.util.ServiceLevelValidator;
+
+import com.google.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.xnap.commons.i18n.I18n;
+
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * ActivationKeyResource
+ */
+public class ActivationKeyController {
+
+    private static final Logger log = LoggerFactory.getLogger(ActivationKeyController.class);
+    private static final Pattern AK_CHAR_FILTER = Pattern.compile("^[a-zA-Z0-9_-]+$");
+
+    private final ActivationKeyCurator activationKeyCurator;
+    private final OwnerProductCurator ownerProductCurator;
+    private final PoolManager poolManager;
+    private final I18n i18n;
+    private final ServiceLevelValidator serviceLevelValidator;
+    private final ActivationKeyRules activationKeyRules;
+
+    @Inject
+    public ActivationKeyController(ActivationKeyCurator activationKeyCurator, I18n i18n,
+        PoolManager poolManager, ServiceLevelValidator serviceLevelValidator,
+        ActivationKeyRules activationKeyRules, OwnerProductCurator ownerProductCurator) {
+
+        this.activationKeyCurator = Objects.requireNonNull(activationKeyCurator);
+        this.i18n = Objects.requireNonNull(i18n);
+        this.poolManager = Objects.requireNonNull(poolManager);
+        this.serviceLevelValidator = Objects.requireNonNull(serviceLevelValidator);
+        this.activationKeyRules = Objects.requireNonNull(activationKeyRules);
+        this.ownerProductCurator = Objects.requireNonNull(ownerProductCurator);
+    }
+
+    /**
+     * Fetches an activation key using the specified key ID. If a valid activation key could not be
+     * found, this method throws an exception.
+     *
+     * @param keyId
+     *  The ID of the activation key to fetch
+     *
+     * @throws BadRequestException
+     *  if the given ID is null, empty or is not associated with a valid activation key
+     *
+     * @return
+     *  an ActivationKey with the specified ID
+     */
+    private ActivationKey fetchActivationKey(String keyId) {
+        if (keyId == null || keyId.isEmpty()) {
+            throw new BadRequestException(i18n.tr("activation key ID is null or empty"));
+        }
+
+        ActivationKey key = this.activationKeyCurator.secureGet(keyId);
+
+        if (key == null) {
+            throw new BadRequestException(i18n.tr("ActivationKey with id {0} could not be found.", keyId));
```

Comment:
The controllers shouldn't be throwing exceptions like this; that's up to the resource to translate.